### PR TITLE
Flake eval cache fixes

### PR DESCRIPTION
**Pull request by roberth**
## Motivation

Get fixes into the release.

This fixes a problem where the third instead of second eval come from cache.
I haven't reproduced the string context splitting problem outside of a unit test in my WIP eval cache work - I would prefer to spend time improving that work than producing a test case for this unimportant corner case.

## Context

- Part of STF "Investigate evaluator performance improvements"

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/eval-cache.cc [hunk 1]
```diff
@@ -124,13 +135,23 @@ struct AttrDb
...
+            // Use UPDATE to preserve rowid if entry exists (prevents orphaning children).
+            auto queryAttribute(state->queryAttribute.use()(key.first)(symbols[key.second]));
+            AttrId rowId;
+            if (queryAttribute.next()) {
+                rowId = (AttrId) queryAttribute.getInt(0);
+                state->updateAttribute.use()(AttrType::FullAttrs)(0, false)(key.first)(symbols[key.second]).exec();
+            } else {
+                state->insertAttribute.use()(key.first)(symbols[key.second])(AttrType::FullAttrs) (0, false).exec();
+                rowId = state->db.getLastInsertedRowId();
+            }
```
Comment (edolstra) [inline src/libexpr/eval-cache.cc]: What's the performance impact of this (since it doubles the number of SQLite calls)?

Maybe you could use `INSERT ... ON CONFLICT DO UPDATE` to do it in a single SQLite call that should preserve the rowid (unlike `INSERT OR REPLACE`).

Comment (roberth) [reply src/libexpr/eval-cache.cc]: [Done](https://github.com/NixOS/nix/compare/2af7f1e191cb7b98f1cf34e02e7339cc2cc9799e..06b9384b3aa81e0e12162643ffde2e418713aecc)

**Comment by tomberek** [comment]
Rebase and merge?

**Review by pkpbynum — APPROVED** [review]
👍 on this PR, was looking into `nix develop` cache misses just now and it's fixed precisely by [f7d66e6](https://github.com/NixOS/nix/pull/15288/commits/f7d66e6e912d9bdaf869650887545a1598e3a7fa)

testing:
```
# Stable nix
% rm -rf ~/.cache/nix/eval-cache-v6
% nix --version
nix (Nix) 2.34.8
% nix develop -c true
nix develop -c true  1.86s user 0.67s system 68% cpu 3.684 total
% time nix develop -c true
nix develop -c true  1.36s user 0.55s system 62% cpu 3.039 total
% time nix develop -c true
nix develop -c true  0.17s user 0.19s system 44% cpu 0.809 total

# This branch
% rm -rf ~/.cache/nix/eval-cache-v6
% ./result/bin/nix --version
nix (Nix) 2.36.0pre20260720_dirty
% git diff
% time ./result/bin/nix develop -c true
./result/bin/nix develop -c true  2.20s user 0.82s system 69% cpu 4.359 total
% time ./result/bin/nix develop -c true
./result/bin/nix develop -c true  0.16s user 0.19s system 45% cpu 0.783 total
```

**Review by edolstra — APPROVED** [review]

**Comment by edolstra** [comment]
Sorry for not looking into this sooner. Thanks, I've resolved the merge conflict!
